caas/kubernetes/cloud/cloud.go

```diff
@@ -0,0 +1,148 @@
+// Copyright 2021 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package cloud
+
+import (
+	"io"
+	"io/ioutil"
+
+	"github.com/juju/errors"
+	"k8s.io/client-go/tools/clientcmd"
+	clientcmdapi "k8s.io/client-go/tools/clientcmd/api"
+
+	"github.com/juju/juju/caas/kubernetes/provider/constants"
+	"github.com/juju/juju/cloud"
+)
+
+type CloudParamaters struct {
+	Name            string
+	Description     string
+	HostCloudRegion string
+	Regions         []cloud.Region
+}
+
+// buildCloudFromCluster finishes building a cloud from the provided kube
+// cluster
+func buildCloudFromCluster(c *cloud.Cloud, cluster *clientcmdapi.Cluster) error {
+	c.Endpoint = cluster.Server
+	c.SkipTLSVerify = cluster.InsecureSkipTLSVerify
+	c.AuthTypes = SupportedAuthTypes()
+
+	clusterCAData, err := dataOrFile(cluster.CertificateAuthorityData, cluster.CertificateAuthority)
+	if err != nil {
+		return errors.Annotate(err, "getting cluster CA data")
+	}
+	c.CACertificates = []string{string(clusterCAData)}
+
+	return nil
+}
+
+// ConfigFromReader does the heavy lifting of transforming a reader object into
+// a kubernetes api config
+func ConfigFromReader(reader io.Reader) (*clientcmdapi.Config, error) {
```

Comment:
i can't see tests for these various public functions

Reply:
Done & Done